File: 0847-shortest-path-visiting-all-nodes/0847-shortest-path-visiting-all-nodes.cpp
class Solution {
public:
    int shortestPathLength(vector<vector<int>>& graph) {
        if (graph.size() == 1) return 0;
        
        int n = graph.size(), endingMask = (1 << n) - 1, steps = 0;
        vector<vector<bool>> seen(n, vector<bool>(endingMask));
        vector<pair<int, int>> queue;
        
        for (int i = 0; i < n; ++i){
            queue.push_back(make_pair(i, 1 << i));
            seen[i][1 << i] = true;
        }
        
        while (!queue.empty()){
            vector<pair<int, int>> nextQueue;
            
            for (int i = 0; i < queue.size(); ++i){
                auto[node, mask] = queue[i];
                
                for (int neighbor : graph[node]){
                    int nextMask = mask | (1 << neighbor);
                    if (nextMask == endingMask) return 1 + steps;
                    
                    if (!seen[neighbor][nextMask]){
                        seen[neighbor][nextMask] = true;
                        nextQueue.push_back(make_pair(neighbor, nextMask));
                    }
                }
            }
            
            ++steps;
            queue = nextQueue;
        }
        
        return -1;
    }
};

File: 0847-shortest-path-visiting-all-nodes/0847-shortest-path-visiting-all-nodes.py
class Solution:
    def shortestPathLength(self, graph: List[List[int]]) -> int:
        def dp(node: int, mask: int) -> int:
            state = (node, mask)
            if state in cache:
                return cache[state]
            
            if mask & (mask - 1) == 0:
                # base case mask only has a single "1" meaning that
                # only one node has been visited (the current node)
                return 0
            
            cache[state] = float('inf') # avoid infinite loop in recursion
            for neighbor in graph[node]:
                if mask & (1 << neighbor):
                    already_visited = 1 + dp(neighbor, mask)
                    not_visited = 1 + dp(neighbor, mask ^ (1 << node))
                    cache[state] = min(cache[state], already_visited, not_visited)
            
            return cache[state]
        
        
        n = len(graph)
        ending_mask = (1 << n) - 1
        cache = dict()
        
        return min(dp(node, ending_mask) for node in range(n))